_posts/2016-01-15-locations-organizations-rbac.md

```diff
@@ -0,0 +1,202 @@
+centers. Between these eight offices and three data centers we have about six
+organizations and about 20 products made available to our members. Each of these
+organizations have developers and system administrators that help manage and
+maintain their servers. For this reason we needed to build Foreman in a way
+where everyone can manage their servers and keep Enterprise Technologies from
+being a bottleneck.
+
+###Locations and Organizations
+
+Locations was a no brainer for us. With eight offices and three data centers we
+definitely needed a way to classify the location of our hosts and in turn use
+that classification to help configure our hosts with certain location specific
+data. Enabling Locations in Foreman solved this for us since the location
+parameter can be used as part of Hiera. The main caveat here is that each host
+has to have the location assigned correctly, but when you are provisioning via
+Foreman and have all of your filters configured you will learn pretty quickly
+if you have the wrong location selected. For existing hosts that we were
+migrating into our new deployment we wrote a custom Fact that sets the host
+location based on the subnet. We simply then changed the setting in Foreman to
+look at the custom Fact for setting the location.
+
+Hiera Location Example:
+
+```
+[root@devbrain ~]# cd /opt/vagrant/puppet/hieradata/
+WARNING Hieradata is a git repo! Check Your Branch and Pull if Needed
+[root@devbrain hieradata]# ls
+application  common.eyaml  domain  environment  location  node  organization  osfamily  security_zone  vtl
+[root@devbrain hieradata]# cd location/
+[root@devbrain location]# ls
+ashburn.eyaml  austin.eyaml  aws.eyaml
+[root@devbrain location]# cat ashburn.eyaml
+---
+# resolv.conf data
+resolv_conf::nameservers:
+  - "dns server ip address"
+  - "dns server ip address"
+
+# ntp configuration data
+ntp::autoupdate: false
+ntp::service_enable: true
+ntp::servers:
+  - "ntp server ip address"
+  - "ntp server ip address"
+
+# repository server data
+repos::repoweb: repomirror.domain.com
+
+# timezone data
+timezone::timezone: EST5EDT
+
+# newrelic data
+newrelic::nrsysmond_labels:
+  - location:ashburn
+
+# nrpe data
+nrpe::allowed_hosts:
+  - '127.0.0.1'
+  - 'nagiosxi.domain.com'
+[root@devbrain location]#
+```
+
+Location Fact Source:
+
+```ruby
+require 'ipaddr'
+ashburn = [
+  IPAddr.new("subnet/16"),
+  IPAddr.new("subnet/23"),
+  IPAddr.new("subnet/23"),
+]
+
+austin = [
+  IPAddr.new("subnet/16"),
+  IPAddr.new("subnet/23"),
+  IPAddr.new("subnet/23"),
+  IPAddr.new("subnet/24"),
+]
+
+aws = [
+  IPAddr.new("subnet/16"),
+]
+
+Facter.add("location") do
+  setcode do
+    network = Facter.value(:ipaddress)
+
+    case
+      when ashburn.any? { |i| i.include?(network)}
+        'ashburn'
+      when austin.any? { |i| i.include?(network)}
+	'austin'
+      when aws.any? { |i| i.include?(network)}
+        'aws'
+      else
+	'unknown'
+    end
+  end
+end
+```
```

Comment:
@GregSutcliffe line 36 should include that we changed the setting to look at the custom fact rather than the default $foreman_location. I can make it more clear though.